fedora-ostree-pruner/fedora-ostree-pruner

```diff
@@ -0,0 +1,267 @@
+    # that should be run
+    if policy.startswith('depth:'):
+        number = policy[6:]
+        return f'--depth={number}'
+    if policy.startswith('time:'):
+        number = policy[5:]
+        return f'--keep-younger-than={number} days ago'
+
+
+def delete_ref_in_repo(ref: str, repo: str, test: bool):
+    logger.info('Deleting the %s ref in repo %s.' % (ref, repo))
+    if test:
+        logger.info('Skipping delete because of test mode')
+        return
+
+    # Delete the ref. We don't worry about pruning here as we'll
+    # run a generic prune for unreachable objects afterwards
+    cmd = ['ostree', 'refs', '--repo', repo, '--delete', ref]
+    runcmd(cmd, capture_output=False)
+
+    # Update the summary file since we deleted a ref
+    cmd = ['ostree', 'summary', '--repo', repo, '-u']
+    runcmd(cmd, capture_output=False)
+
+
+def prune_compose_repo(test=False):
+    # prune the compose repo
+    logger.info('Pruning the compose repo %s with policy %s' %
+                (OSTREECOMPOSEREPO, COMPOSE_REPO_POLICY))
+    prunearg = parse_policy(COMPOSE_REPO_POLICY)
+    cmd = ['ostree', 'prune', '--repo', OSTREECOMPOSEREPO,
+           '--refs-only', prunearg]
+    if test:
+        cmd.append('--no-prune')
+    runcmd(cmd, capture_output=False)
+
+
+def prune_prod_repo(test=False):
+    # Error out if any refs exist that aren't defined in the policy
+    cmd = ['ostree', 'refs', '--repo', OSTREEPRODREPO]
+    cp = runcmd(cmd)
+    prod_refs = cp.stdout.decode('utf-8').splitlines()
+    for ref in prod_refs:
+        if ref not in PROD_REF_POLICIES:
+            msg = 'ref %s in repo %s but no policy defined' % (ref, OSTREEPRODREPO)
+            logger.error(msg)
+            raise Exception(msg)
+
+    # Warn up front for refs that are in the policy but not in the repo
+    for ref,policy in PROD_REF_POLICIES.items():
+        if ref not in prod_refs:
+            if policy != 'delete':
+                logger.warning('policy defined for a ref that is not in the repo %s: %s' %
+                               (OSTREEPRODREPO, ref))
+            continue
+
+    # prune each branch in the policy with specified value
+    deleted_a_ref = False
+    for ref,policy in PROD_REF_POLICIES.items():
+
+        # Skip if there is no ref in the repo for this policy
+        if ref not in prod_refs:
+            continue
+
+        if policy is None:
+            logger.info('Skipping ref %s in repo %s. Policy is to keep all commits' %
+                        (ref, OSTREEPRODREPO))
+            continue
+
+        if policy == 'delete':
+            delete_ref_in_repo(ref, OSTREEPRODREPO, test)
+            deleted_a_ref = True
+            continue
+
+        prunearg = parse_policy(policy)
+
+        logger.info('Pruning the %s ref in repo %s to %s' %
+                    (ref, OSTREEPRODREPO, policy))
+        cmd = ['ostree', 'prune', '--repo', OSTREEPRODREPO,
+               '--only-branch', ref, '--refs-only', prunearg]
```

Comment:
Yes, it would be an RFE. We can start with this first and then see if we really want this. I'm interested to see timing results after the first initial mass pruning.

Reply:
Wanted to give an update on `ostree prune --commit-only`. The feature is in the newest [release of ostree](https://github.com/ostreedev/ostree/releases/tag/v2022.2), so we can proceed with the `fedora-ostree-pruner` work.